linera-views/src/common.rs

```diff
@@ -122,6 +125,121 @@ where
     }
 }
 
+/// GreatestLowerBoundIterator iterates over the entries of a BTreeSet.
+/// The function call `get_lower_bound(val)` returns a `Some(x)` where `x` is the highest
+/// entry such that `x <= val`. If none exists then None is returned.
+///
+/// The function calls `is_index_absent` have to be done with increasing
+/// values in order to get correct results.
+pub(crate) struct GreatestLowerBoundIterator<'a, IT> {
+    prefix_len: usize,
+    prec1: Option<&'a Vec<u8>>,
+    prec2: Option<&'a Vec<u8>>,
+    iter: IT,
+}
+
+impl<'a, IT> GreatestLowerBoundIterator<'a, IT>
+where
+    IT: Iterator<Item = &'a Vec<u8>>,
+{
+    pub(crate) fn new(prefix_len: usize, mut iter: IT) -> Self {
+        let prec1 = None;
+        let prec2 = iter.next();
+        Self {
+            prefix_len,
+            prec1,
+            prec2,
+            iter,
+        }
+    }
+
+    fn compar(&self, x: &'a Vec<u8>, val: &[u8]) -> bool {
+        let len1 = x.len() - self.prefix_len;
+        let len2 = val.len();
+        let min_len = len1.min(len2);
+        for u in 0..min_len {
+            if x[self.prefix_len + u] > val[u] {
+                return true;
+            }
+        }
+        len1 > len2
+    }
+
+    fn get_lower_bound(&mut self, val: &[u8]) -> Option<&'a Vec<u8>> {
+        loop {
+            match &self.prec2 {
+                None => {
+                    return self.prec1;
+                }
+                Some(x) => {
+                    if self.compar(x, val) {
```

Comment:
Isn't that just the default ordering?

```suggestion
                    if &x[prefix.len()..] > val {
```

Reply:
Yes, you are right. Sorry for that. I was searching around for some time and somehow I missed that simple solution.